src/Agent.Listener/Configuration/ConfigurationProvider.cs

```diff
@@ -0,0 +1,357 @@
+﻿using Microsoft.TeamFoundation.DistributedTask.WebApi;
+using Microsoft.VisualStudio.Services.Agent.Util;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.TeamFoundation.Common;
+using Microsoft.VisualStudio.Services.Common;
+using Microsoft.VisualStudio.Services.WebApi;
+
+namespace Microsoft.VisualStudio.Services.Agent.Listener.Configuration
+{
+    public interface IConfigurationProvider : IExtension
+    {
+        void InitConnection(IAgentServer agentServer);
+
+        string ConfigurationProviderType { get; }
+
+        string GetServerUrl(CommandSettings command);
+
+        Task<IAgentServer> TestConnectAsync(string tfsUrl, VssCredentials creds);
+
+        Task<int> GetPoolId(CommandSettings command);
+
+        Task<TaskAgent> UpdateAgentAsync(int poolId, TaskAgent agent);
+
+        Task<TaskAgent> AddAgentAsync(int poolId, TaskAgent agent);
+
+        Task DeleteAgentAsync(int agentPoolId, int agentId);
+
+        void UpdateAgentSetting(AgentSettings settings);
+    }
+
+    public abstract class ConfigurationProvider : AgentService
+    {
+        public Type ExtensionType => typeof(IConfigurationProvider);
+        protected ITerminal _term;
+        protected IAgentServer _agentServer;
+
+        public override void Initialize(IHostContext hostContext)
+        {
+            base.Initialize(hostContext);
+            _term = hostContext.GetService<ITerminal>();
+        }
+
+        protected void InitializeServerConnection(IAgentServer agentServer)
+        {
+            _agentServer = agentServer;
```

Comment:
this is against the purpose of ServiceLocator pattern we have. any time you need the agentserver, it should return from HostContext.GetService<IAgentServer>()